src/main/java/com/o19s/es/explore/ExplorerQueryBuilder.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.o19s.es.explore;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.NamedWriteable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+
+public class ExplorerQueryBuilder extends AbstractQueryBuilder<ExplorerQueryBuilder> implements NamedWriteable{
+    public static final String NAME = "match_explorer";
+
+    private static final ParseField QUERY_NAME = new ParseField("query");
+    private static final ParseField TYPE_NAME = new ParseField("type");
+
+    private QueryBuilder query;
+    private String type;
+
+    private static final ObjectParser<ExplorerQueryBuilder, QueryParseContext> PARSER;
+
+    static {
+        PARSER = new ObjectParser<>(NAME, ExplorerQueryBuilder::new);
+        PARSER.declareObject(
+                ExplorerQueryBuilder::query,
+                (parser, context) -> context.parseInnerQueryBuilder().get(),
+                QUERY_NAME
+        );
+        PARSER.declareString(ExplorerQueryBuilder::statsType, TYPE_NAME);
+        declareStandardFields(PARSER);
+    }
+
+
+    public ExplorerQueryBuilder() {
+    }
+
+    public ExplorerQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readNamedWriteable(QueryBuilder.class);
+        type = in.readString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(query);
+        out.writeString(type);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+
+        printBoostAndQueryName(builder);
+
+        builder.field(QUERY_NAME.getPreferredName());
```

Comment:
nitpick: you can pack these 2 lines into one:
`builder.field(QUERY_NAME.getPreferredName(), query);`